File: core/core/src/types/copy.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use std::pin::Pin;
use std::task::Context;
use std::task::Poll;
use std::task::ready;
use std::{fmt, fmt::Formatter};

use futures::Stream;

use crate::raw::*;
use crate::*;

/// Copier is designed to drive long-running copy operations.
///
/// - Copier implements `Stream<Item = Result<usize>>`.
/// - `Some(Ok(n))` means the copy operation made progress by `n` bytes.
/// - `None` means the copy operation has completed.
pub struct Copier {
    copier: Option<oio::Copier>,
    metadata: Option<Metadata>,

    fut: Option<BoxedStaticFuture<(oio::Copier, Result<Option<usize>>)>>,
    errored: bool,
}

impl fmt::Debug for Copier {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        f.debug_struct("Copier")
            .field("active", &self.copier.is_some())
            .field("errored", &self.errored)
            .finish()
    }
}

/// # Safety
///
/// Copier will only be accessed by `&mut Self`.
unsafe impl Sync for Copier {}

impl Copier {
    /// Create a new copier.
    pub(crate) async fn create(
        acc: Accessor,
        from: &str,
        to: &str,
        args: OpCopy,
        opts: OpCopier,
    ) -> Result<Self> {
        let (_, copier) = acc.copy(from, to, args, opts).await?;

        Ok(Self {
            copier: Some(copier),
            metadata: None,

            fut: None,
            errored: false,
        })
    }

    /// Drive the copy operation forward.
    pub async fn next(&mut self) -> Result<Option<usize>> {
        if self.errored {
            return Ok(None);
        }

        let Some(copier) = self.copier.as_mut() else {
            return Ok(None);
        };

        match copier.next_dyn().await {
            Ok(Some(n)) => Ok(Some(n)),
            Ok(None) => {
                self.metadata = Some(copier.close_dyn().await?);
                self.copier = None;
                Ok(None)
            }
            Err(err) => {
                self.errored = true;
                Err(err)
            }
        }
    }

    /// Abort the pending copy operation.
    pub async fn abort(&mut self) -> Result<()> {
        let Some(copier) = self.copier.as_mut() else {
            return Ok(());
        };

        copier.abort_dyn().await
    }

    /// Close the copier and return metadata from the server-side completion response.
    pub async fn close(&mut self) -> Result<Metadata> {
        if self.errored {
            return Err(Error::new(
                ErrorKind::Unexpected,
                "copier has already failed and can't be closed",
            ));
        }

        if let Some(metadata) = self.metadata.clone() {
            return Ok(metadata);
        }

        while self.metadata.is_none() {
            match self.next().await? {
                Some(_) => continue,
                None => break,
            }
        }

        Ok(self.metadata.clone().unwrap_or_default())
    }
}

impl Stream for Copier {
    type Item = Result<usize>;

    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        if self.errored {
            return Poll::Ready(None);
        }

        if let Some(mut copier) = self.copier.take() {
            let fut = async move {
                let res = copier.next_dyn().await;
                (copier, res)
            };
            self.fut = Some(Box::pin(fut));
        }

        if let Some(fut) = self.fut.as_mut() {
            let (copier, res) = ready!(fut.as_mut().poll(cx));
            self.copier = Some(copier);
            self.fut = None;

            return match res {
                Ok(Some(n)) => Poll::Ready(Some(Ok(n))),
                Ok(None) => Poll::Ready(None),
                Err(err) => {
                    self.errored = true;
                    Poll::Ready(Some(Err(err)))
                }
            };
        }

        Poll::Ready(None)
    }
}

File: bindings/go/copy.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package opendal

import (
	"context"
	"runtime"
	"unsafe"

	"github.com/jupiterrider/ffi"
)

// WithCopyFn is a functional option for copy operations.
type WithCopyFn func(*copyOptions)

// CopyWithIfNotExists sets whether the copy operation should only succeed
// if the target does not exist.
func CopyWithIfNotExists(ifNotExists bool) WithCopyFn {
	return func(o *copyOptions) {
		o.ifNotExists = ifNotExists
	}
}

// CopyWithIfMatch sets the If-Match condition for the copy operation.
//
// The copy will only succeed when the destination object's ETag matches
// the given value.
func CopyWithIfMatch(ifMatch string) WithCopyFn {
	return func(o *copyOptions) {
		o.ifMatch = ifMatch
	}
}

// CopyWithSourceVersion sets the source version for the copy operation.
//
// The copy will read from the specified source version instead of the
// current source object.
func CopyWithSourceVersion(sourceVersion string) WithCopyFn {
	return func(o *copyOptions) {
		o.sourceVersion = sourceVersion
	}
}

// CopyWithSourceContentLengthHint provides a hint of the source object's
// content length, avoiding extra metadata requests.
func CopyWithSourceContentLengthHint(length uint64) WithCopyFn {
	return func(o *copyOptions) {
		o.sourceContentLengthHint = length
	}
}

// CopyWithConcurrent sets the maximum number of concurrent copy operations.
func CopyWithConcurrent(concurrent uint) WithCopyFn {
	return func(o *copyOptions) {
		o.concurrent = concurrent
	}
}

// CopyWithChunk sets the chunk size for segmented copy operations.
func CopyWithChunk(chunk uint) WithCopyFn {
	return func(o *copyOptions) {
		o.chunk = chunk
	}
}

type copyOptions struct {
	ifNotExists             bool
	ifMatch                 string
	sourceVersion           string
	sourceContentLengthHint uint64
	concurrent              uint
	chunk                   uint
}

// Copy duplicates a file from the source path to the destination path.
//
// Copy is a wrapper around the C-binding function `opendal_operator_copy_with_cancel`.
// When options are provided, it uses `opendal_operator_copy_with_options_cancel`.
//
// # Parameters
//
//   - ctx: The context for the operation. Canceling it cancels the underlying
//     native call in a blocking manner.
//   - src: The source file path.
//   - dest: The destination file path.
//   - opts: Optional copy options.
//
// # Returns
//
//   - error: An error if the copy operation fails, or nil if successful.
//
// # Behavior
//
//   - Both src and dest must be file paths, not directories.
//   - If dest already exists, it will be overwritten.
//   - If src and dest are identical, an IsSameFile error will be returned.
//   - The copy operation is idempotent; repeated calls with the same parameters will yield the same result.
//
// # Example
//
//	func exampleCopy(op *opendal.Operator) {
//		err := op.Copy(context.Background(), "path/from/file", "path/to/file")
//		if err != nil {
//			log.Fatal(err)
//		}
//	}
//
// Note: This example assumes proper error handling and import statements.
func (op *Operator) Copy(ctx context.Context, src, dest string, opts ...WithCopyFn) error {
	if len(opts) == 0 {
		return runErrWithCancelContext(ctx, op.ctx, func(token *opendalCancelToken) error {
			return ffiOperatorCopyWithCancel.symbol(op.ctx)(op.inner, src, dest, token)
		})
	}

	o := &copyOptions{}
	for _, opt := range opts {
		opt(o)
	}
	cOpts := ffiCopyOptionsNew.symbol(op.ctx)()
	free := ffiCopyOptionsFree.symbol(op.ctx)

	ffiCopyOptionsSetIfNotExists.symbol(op.ctx)(cOpts, o.ifNotExists)
	var ifMatchData []byte
	if o.ifMatch != "" {
		data, err := ffiCopyOptionsSetIfMatch.symbol(op.ctx)(cOpts, o.ifMatch)
		if err != nil {
			free(cOpts)
			return err
		}
		ifMatchData = data
	}
	var sourceVersionData []byte
	if o.sourceVersion != "" {
		data, err := ffiCopyOptionsSetSourceVersion.symbol(op.ctx)(cOpts, o.sourceVersion)
		if err != nil {
			free(cOpts)
			return err
		}
		sourceVersionData = data
	}
	if o.sourceContentLengthHint != 0 {
		ffiCopyOptionsSetSourceContentLengthHint.symbol(op.ctx)(cOpts, o.sourceContentLengthHint)
	}
	if o.concurrent != 0 {
		ffiCopyOptionsSetConcurrent.symbol(op.ctx)(cOpts, o.concurrent)
	}
	if o.chunk != 0 {
		ffiCopyOptionsSetChunk.symbol(op.ctx)(cOpts, o.chunk)
	}
	err := runErrWithCancelContext(ctx, op.ctx, func(token *opendalCancelToken) error {
		return ffiOperatorCopyWithOptionsCancel.symbol(op.ctx)(op.inner, src, dest, cOpts, token)
	})
	free(cOpts)
	runtime.KeepAlive(ifMatchData)
	runtime.KeepAlive(sourceVersionData)
	return err
}

var ffiCopyOptionsNew = newFFI(ffiOpts{
	sym:   "opendal_copy_options_new",
	rType: &ffi.TypePointer,
}, func(_ context.Context, ffiCall ffiCall) func() *opendalCopyOptions {
	return func() *opendalCopyOptions {
		var opts *opendalCopyOptions
		ffiCall(unsafe.Pointer(&opts))
		return opts
	}
})

var ffiCopyOptionsFree = newFFI(ffiOpts{
	sym:    "opendal_copy_options_free",
	rType:  &ffi.TypeVoid,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(_ context.Context, ffiCall ffiCall) func(opts *opendalCopyOptions) {
	return func(opts *opendalCopyOptions) {
		ffiCall(
			nil,
			unsafe.Pointer(&opts),
		)
	}
})

var ffiCopyOptionsSetIfNotExists = newFFI(ffiOpts{
	sym:    "opendal_copy_options_set_if_not_exists",
	rType:  &ffi.TypeVoid,
	aTypes: []*ffi.Type{&ffi.TypePointer, &ffi.TypeUint8},
}, func(_ context.Context, ffiCall ffiCall) func(opts *opendalCopyOptions, ifNotExists bool) {
	return func(opts *opendalCopyOptions, ifNotExists bool) {
		var v uint8
		if ifNotExists {
			v = 1
		}
		ffiCall(nil, unsafe.Pointer(&opts), unsafe.Pointer(&v))
	}
})

var ffiCopyOptionsSetIfMatch = newFFI(ffiOpts{
	sym:    "opendal_copy_options_set_if_match",
	rType:  &ffi.TypeVoid,
	aTypes: []*ffi.Type{&ffi.TypePointer, &ffi.TypePointer},
}, func(_ context.Context, ffiCall ffiCall) func(opts *opendalCopyOptions, ifMatch string) ([]byte, error) {
	return func(opts *opendalCopyOptions, ifMatch string) ([]byte, error) {
		data, err := byteSliceFromString(ifMatch)
		if err != nil {
			return nil, err
		}
		byteValue := &data[0]
		ffiCall(nil, unsafe.Pointer(&opts), unsafe.Pointer(&byteValue))
		return data, nil
	}
})

var ffiCopyOptionsSetSourceVersion = newFFI(ffiOpts{
	sym:    "opendal_copy_options_set_source_version",
	rType:  &ffi.TypeVoid,
	aTypes: []*ffi.Type{&ffi.TypePointer, &ffi.TypePointer},
}, func(_ context.Context, ffiCall ffiCall) func(opts *opendalCopyOptions, sourceVersion string) ([]byte, error) {
	return func(opts *opendalCopyOptions, sourceVersion string) ([]byte, error) {
		data, err := byteSliceFromString(sourceVersion)
		if err != nil {
			return nil, err
		}
		byteValue := &data[0]
		ffiCall(nil, unsafe.Pointer(&opts), unsafe.Pointer(&byteValue))
		return data, nil
	}
})

var ffiCopyOptionsSetSourceContentLengthHint = newFFI(ffiOpts{
	sym:    "opendal_copy_options_set_source_content_length_hint",
	rType:  &ffi.TypeVoid,
	aTypes: []*ffi.Type{&ffi.TypePointer, &ffi.TypeUint64},
}, func(_ context.Context, ffiCall ffiCall) func(opts *opendalCopyOptions, sourceContentLengthHint uint64) {
	return func(opts *opendalCopyOptions, sourceContentLengthHint uint64) {
		ffiCall(nil, unsafe.Pointer(&opts), unsafe.Pointer(&sourceContentLengthHint))
	}
})

var ffiCopyOptionsSetConcurrent = newFFI(ffiOpts{
	sym:    "opendal_copy_options_set_concurrent",
	rType:  &ffi.TypeVoid,
	aTypes: []*ffi.Type{&ffi.TypePointer, &ffi.TypePointer},
}, func(_ context.Context, ffiCall ffiCall) func(opts *opendalCopyOptions, concurrent uint) {
	return func(opts *opendalCopyOptions, concurrent uint) {
		c := uintptr(concurrent)
		ffiCall(nil, unsafe.Pointer(&opts), unsafe.Pointer(&c))
	}
})

var ffiCopyOptionsSetChunk = newFFI(ffiOpts{
	sym:    "opendal_copy_options_set_chunk",
	rType:  &ffi.TypeVoid,
	aTypes: []*ffi.Type{&ffi.TypePointer, &ffi.TypePointer},
}, func(_ context.Context, ffiCall ffiCall) func(opts *opendalCopyOptions, chunk uint) {
	return func(opts *opendalCopyOptions, chunk uint) {
		c := uintptr(chunk)
		ffiCall(nil, unsafe.Pointer(&opts), unsafe.Pointer(&c))
	}
})

var ffiOperatorCopyWithCancel = newFFI(ffiOpts{
	sym:    "opendal_operator_copy_with_cancel",
	rType:  &ffi.TypePointer,
	aTypes: []*ffi.Type{&ffi.TypePointer, &ffi.TypePointer, &ffi.TypePointer, &ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(op *opendalOperator, src, dest string, token *opendalCancelToken) error {
	return func(op *opendalOperator, src, dest string, token *opendalCancelToken) error {
		byteSrc, err := BytePtrFromString(src)
		if err != nil {
			return err
		}
		byteDest, err := BytePtrFromString(dest)
		if err != nil {
			return err
		}
		var e *opendalError
		ffiCall(
			unsafe.Pointer(&e),
			unsafe.Pointer(&op),
			unsafe.Pointer(&byteSrc),
			unsafe.Pointer(&byteDest),
			unsafe.Pointer(&token),
		)
		return parseError(ctx, e)
	}
})

var ffiOperatorCopyWithOptionsCancel = newFFI(ffiOpts{
	sym:    "opendal_operator_copy_with_options_cancel",
	rType:  &ffi.TypePointer,
	aTypes: []*ffi.Type{&ffi.TypePointer, &ffi.TypePointer, &ffi.TypePointer, &ffi.TypePointer, &ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(op *opendalOperator, src, dest string, opts *opendalCopyOptions, token *opendalCancelToken) error {
	return func(op *opendalOperator, src, dest string, opts *opendalCopyOptions, token *opendalCancelToken) error {
		byteSrc, err := BytePtrFromString(src)
		if err != nil {
			return err
		}
		byteDest, err := BytePtrFromString(dest)
		if err != nil {
			return err
		}
		var e *opendalError
		ffiCall(
			unsafe.Pointer(&e),
			unsafe.Pointer(&op),
			unsafe.Pointer(&byteSrc),
			unsafe.Pointer(&byteDest),
			unsafe.Pointer(&opts),
			unsafe.Pointer(&token),
		)
		return parseError(ctx, e)
	}
})
